src/f32-spmm/gen/f32-spmm-32x1-minmax-hvx-pipelined-x2.c

```diff
@@ -0,0 +1,263 @@
+// Auto-generated file. Do not edit!
+//   Template: src/f32-spmm/hvx-pipelined.c.in
+//   Generator: tools/xngen
+//
+// Copyright 2024 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <assert.h>
+
+#include <hvx_hexagon_protos.h>
+#include <hexagon_protos.h>
+#include <hexagon_types.h>
+
+#include "xnnpack/simd/f32-hvx.h"
+
+#include "xnnpack/prefetch.h"
+#include "xnnpack/spmm.h"
+
+void xnn_f32_spmm_minmax_ukernel_32x1__hvx_pipelined_x2(
+    size_t mc,
+    size_t nc,
+    const float* input,
+    const float* weights,
+    const int32_t* widx_dmap,
+    const uint32_t* nidx_nnzmap,
+    float* output,
+    size_t output_stride,
+    const union xnn_f32_minmax_params params[restrict XNN_MIN_ELEMENTS(1)])
+{
+  assert(mc != 0);
+  assert(mc % sizeof(float) == 0);
+  assert(nc != 0);
+
+  const HVX_Vector vmin = xnn_set1_f32(params->scalar.min);
+  const HVX_Vector vmax = xnn_set1_f32(params->scalar.max);
+
+  size_t output_decrement = output_stride * nc - 32 * sizeof(float);
+  while XNN_LIKELY(mc >= 32 * sizeof(float)) {
+    const float* w = weights;
+    const int32_t* dmap = widx_dmap;
+    const uint32_t* nnzmap = nidx_nnzmap;
+    HVX_Vector vw = xnn_set1_f32(*w); w += 1;
+
+    intptr_t diff = *dmap++;
+
+    HVX_Vector vi0 = xnn_loadu_f32(input);
+
+    size_t n = nc;
+
+    do {
+      uint32_t nnz = *nnzmap++;
+      HVX_Vector vacc0 = vw;
+      vw = xnn_set1_f32(*w); w += 1;
+
+      for(; nnz >= 2; nnz -= 2) {
+        vacc0 = xnn_fmadd_f32(vi0, vw, vacc0);
+        input = (const float*) ((uintptr_t) input + (uintptr_t) diff);
+        xnn_prefetch_to_l1(input + 128);
+
+        diff = *dmap++;
+
+        vw = xnn_set1_f32(*w); w += 1;
+        xnn_prefetch_to_l1(w + 32);
+
+        vi0 = xnn_loadu_f32(input + 0);
+        vacc0 = xnn_fmadd_f32(vi0, vw, vacc0);
+        input = (const float*) ((uintptr_t) input + (uintptr_t) diff);
+        xnn_prefetch_to_l1(input + 128);
+
+        diff = *dmap++;
+
+        vw = xnn_set1_f32(*w); w += 1;
+        xnn_prefetch_to_l1(w + 32);
+
+        vi0 = xnn_loadu_f32(input + 0);
+      }
+      if XNN_LIKELY(nnz != 0) {
+        do {
+          vacc0 = xnn_fmadd_f32(vi0, vw, vacc0);
+          input = (const float*) ((uintptr_t) input + (uintptr_t) diff);
+          xnn_prefetch_to_l1(input + 128);
```

Comment:
I think on Hexagon, vectors only go to L2? This might still be a useful way to prefetch, but perhaps we need to add a `xnn_prefetch_to_l2` that maps to Hexagon's l2fetch.

This should probably wait until we can dig into performance on real hardware more thoroughly.

Reply:
I revisited the manual to confirm and I found the correct way to use `dcfetch` is - (1) bring data into L2 first, then (2) bring data into L1 from L2. This is the description from the manual - 

"It is common for algorithms to use the vector engine to produce results that must further process on the scalar core. Use VMEM stores to get the data into L2, then use DCFETCH to get the data in L1, followed by scalar load instructions. Execute the DCFETCH anytime after the VMEM store, however, software should budget at least 30 cycles before issuing 
the scalar load instruction".

Probably I can change xnn_prefetch_to_l1 to have:
```
xnn_prefetch_to_l2(addr, size)
dcfetch(addr)
```
and add `xnn_prefetch_to_l2` to call `l2fetch(addr, size)`. And agree with you about performance analysis first, and we can improve this. Running on devices is one of my active items using Steven's script.

Reply:
I think we should add an `xnn_prefetch_to_l2` function. But IIRC, Hexagon l2fetches can be 2D, so I think the arguments should be something like: `xnn_prefetch_to_l2_2d(addr, width_bytes, height, stride)` or similar. Maybe we should also have an overload like you described that assumes height = 1 (and then stride is irrelevant).

Reply:
I just added `xnn_prefetch_to_l2` in `prefetch.h`. The format is in `xnn_prefetch_to_l2(address, stride, width, height, iter)` - this calls `l2fetch(addr, info)` where `info` has the all the information about stride, width, height and iteration. I got hints from Halide how to prepare the `info`.

When we optimize this code later, we can check - if prefetch is called at the right timing and with right amount (the manual says at least few hundreds cycle before using data).

Reply:
Changed slightly having prefetch_to_l2_1D and 2D. In this code, we use 1D function so it is more clear than using 2D function like a overload.